src/jit/emitxarch.cpp

```diff
@@ -3063,88 +3081,68 @@ regNumber emitter::emitInsBinary(instruction ins, emitAttr attr, GenTree* dst, G
     return result;
 }
 
-/** Emit logic for Read-Modify-Write Instructions
- *  Responsible for encoding a single instruction that will perform an operation of the form
- *  *addr op= operand for example
- *  ADD [RAX], RCX
- *
- *  Preconditions:  Lowering has taken care of recognizing the StoreInd pattern of
- *                  StoreInd( AddressTree, BinOp( Ind ( AddressTree ), Operand )
- *                  The address to store is already sitting in a register.
- *
- *                  This is a no-produce operation, meaning that no register output will
- *                  be produced for future use in the code stream.
- *
- * The caller is responsible for calling genConsumeReg() on all source registers, and
- * genProduceReg on the target register, if any.
- */
-void emitter::emitInsRMW(instruction ins, emitAttr attr, GenTree* dstAddr, GenTree* src)
+//------------------------------------------------------------------------
+// emitInsRMW: Emit logic for Read-Modify-Write binary instructions.
+//
+// Responsible for emitting a single instruction that will perform an operation of the form:
+//      *addr = *addr <BinOp> src
+// For example:
+//      ADD [RAX], RCX
+//
+// Arguments:
+//    ins - instruction to generate
+//    attr - emitter attribute for instruction
+//    storeInd - indir for RMW addressing mode
+//    src - source operand of instruction
+//
+// Assumptions:
+//    Lowering has taken care of recognizing the StoreInd pattern of:
+//          StoreInd( AddressTree, BinOp( Ind ( AddressTree ), Operand ) )
+//    The address to store is already sitting in a register.
+//
+// Notes:
+//    This is a no-produce operation, meaning that no register output will
+//    be produced for future use in the code stream.
+//
+//    The caller is responsible for calling genConsumeReg() on all source registers, and
+//    genProduceReg on the target register, if any.
+//
+void emitter::emitInsRMW(instruction ins, emitAttr attr, GenTreeStoreInd* storeInd, GenTree* src)
 {
-    assert(dstAddr->isIndir());
-    GenTreeIndir* indir = dstAddr->AsIndir();
-    GenTreePtr mem = indir->Addr();
-    assert(mem->gtSkipReloadOrCopy()->OperGet() == GT_LCL_VAR ||
-           mem->gtSkipReloadOrCopy()->OperGet() == GT_LCL_VAR_ADDR ||
-           mem->gtSkipReloadOrCopy()->OperGet() == GT_LEA ||
-           mem->gtSkipReloadOrCopy()->OperGet() == GT_CLS_VAR_ADDR ||
-           mem->gtSkipReloadOrCopy()->OperGet() == GT_CNS_INT);
+    GenTreePtr mem = storeInd->Addr();
+    mem = mem->gtSkipReloadOrCopy();
+    assert(mem->OperGet() == GT_LCL_VAR ||
```

Comment:
Not a big deal - but "mem" is weird to me because it is somewhat ambiguous whether you are referring to the actual memory contents (which is usually represented by a GT_IND), or the address of the memory. So "addr" seems better to me, but since it was already "mem" it is fine to leave as-is.